filters/include/bloom_filter.hpp

```diff
@@ -0,0 +1,757 @@
+   * @param print_filter If true, the filter bits will be printed as well.
+   * @return A human-readable string representation of the Bloom Filter.
+   */
+  string<A> to_string(bool print_filter = false) const;
+
+private:
+  using AllocUint8 = typename std::allocator_traits<A>::template rebind_alloc<uint8_t>;
+
+  static const uint64_t DIRTY_BITS_VALUE = static_cast<uint64_t>(-1LL);
+  static const uint64_t MAX_HEADER_SIZE_BYTES = 32; // 4 Java Longs
+  static const uint64_t BIT_ARRAY_LENGTH_OFFSET_BYTES = 16;
+  static const uint64_t NUM_BITS_SET_OFFSET_BYTES = 24;
+  static const uint64_t BIT_ARRAY_OFFSET_BYTES = 32;
+  static const uint64_t MAX_FILTER_SIZE_BITS = (INT32_MAX - MAX_HEADER_SIZE_BYTES) * sizeof(uint64_t);
+
+  static const uint8_t PREAMBLE_LONGS_EMPTY = 3;
+  static const uint8_t PREAMBLE_LONGS_STANDARD = 4;
+  static const uint8_t FAMILY_ID = 21;
+  static const uint8_t SER_VER = 1;
+  static const uint8_t EMPTY_FLAG_MASK = 4;
+
+  // used by builder methods
+  bloom_filter_alloc(const uint64_t num_bits, const uint16_t num_hashes, const uint64_t seed, const A& allocator);
+  bloom_filter_alloc(uint8_t* memory, size_t length_bytes, const uint64_t num_bits, const uint16_t num_hashes, const uint64_t seed, const A& allocator);
+
+  // used by deserialize and wrap
+  bloom_filter_alloc(const uint64_t seed,
+                     const uint16_t num_hashes,
+                     const bool is_dirty,
+                     const bool is_owned,
+                     const bool is_read_only,
+                     const uint64_t capacity_bits,
+                     const uint64_t num_bits_set,
+                     uint8_t* bit_array,
+                     uint8_t* memory,
+                     const A& allocator);
+
+  static bloom_filter_alloc internal_deserialize_or_wrap(void* bytes,
+                                                         size_t length_bytes,
+                                                         bool read_only,
+                                                         bool wrap,
+                                                         const A& allocator);
+
+  // internal query/update methods
+  void internal_update(const uint64_t h0, const uint64_t h1);
+  bool internal_query_and_update(const uint64_t h0, const uint64_t h1);
+  bool internal_query(const uint64_t h0, const uint64_t h1) const;
+
+  void update_num_bits_set(const uint64_t num_bits_set);
+
+  A allocator_;
+  uint64_t seed_;
+  uint16_t num_hashes_;
+  bool is_dirty_;
+  bool is_owned_; // if true, data is not owned by filter AND data_ holdes the entire filter not just the bit array
+  bool is_read_only_; // if true, filter is read-only
+  uint64_t capacity_bits_;
+  uint64_t num_bits_set_;
+  uint8_t* bit_array_;  // data backing bit_array_, regardless of ownership
+  uint8_t* memory_; // if wrapped, pointer to the start of the filter, otheriwse nullptr
+
+  friend class bloom_filter_builder_alloc<A>;
```

Comment:
don't we want to declare the builder in this class as we do in Theta sketch?

Reply:
Agnostic on this, so made it parallel to theta for consistency